src/map/pc.c

```diff
@@ -1624,33 +1624,30 @@ static int pc_calc_skilltree(struct map_session_data *sd)
 			sd->status.skill[i].lv = (sd->status.skill[i].flag == SKILL_FLAG_TEMPORARY) ? 0 : sd->status.skill[i].flag - SKILL_FLAG_REPLACED_LV_0;
 			sd->status.skill[i].flag = SKILL_FLAG_PERMANENT;
 		}
+	}
 
-		if (sd->sc.count && sd->sc.data[SC_SOULLINK] && sd->sc.data[SC_SOULLINK]->val2 == SL_BARDDANCER
-		 && ((skill->dbs->db[i].nameid >= BA_WHISTLE && skill->dbs->db[i].nameid <= BA_APPLEIDUN)
-		  || (skill->dbs->db[i].nameid >= DC_HUMMING && skill->dbs->db[i].nameid <= DC_SERVICEFORYOU))
-		) {
-			//Enable Bard/Dancer spirit linked skills.
-			int linked_nameid = skill->get_linked_song_dance_id(skill->dbs->db[i].nameid);
-			if (linked_nameid == 0) {
-				Assert_report("Linked bard/dance skill not found");
-				continue;
+	//Enable Bard/Dancer spirit linked skills.
+	if (sd->sc.count && sd->sc.data[SC_SOULLINK] && sd->sc.data[SC_SOULLINK]->val2 == SL_BARDDANCER) {
+		int bard_song_skillidx[4] = {skill->get_index(BA_WHISTLE), skill->get_index(BA_ASSASSINCROSS), skill->get_index(BA_POEMBRAGI), skill->get_index(BA_APPLEIDUN)};
```

Comment:
Split line, or more all skill->get_index to the outside and do them once u need the index.
Add const qualifier to the array.

You could add sth like this if you rename the the arrays accordingly:
```c
		const int *from_nameid = NULL;
		const int *to_nameid = NULL;
		if (sd->status.sex == SEX_MALE) {
			from_nameid = bard_skills;
			to_nameid = dancer_skills;
		} else {
			from_nameid = dancer_skills;
			to_nameid = bard_skills;
		}
```

Also add this
```c
STATIC_ASSERT(ARRAYLENGTH(bard_skills) == ARRAYLENGTH(dancer_skills), "bard_skills and dancer_skills must be the same size");
```

Then you can loop over ARRAYLENGHT(bard_skills)
and inside the loop you use that if that you have accordingly but before you do sth like this:
```c
			int copy_from_index = skill->get_index(from_nameid[i]);
			int copy_to_index = skill->get_index(to_nameid[i]);
			if (copy_from_index == 0 || copy_to_index == 0) {
				Assert_report("Linked bard/dancer skill not found");
				continue;
			}
```
Which is also nice because you add more safety in case a skill->get_index() fails.

Credits to this solution go to Haru btw.